crates/wysiwyg/src/dom/parser/parse.rs

```diff
@@ -12,256 +12,513 @@
+    fn to_dom_creation_error<S, E>(error: E) -> DomCreationError<S>
     where
-        T: AsRef<str>,
+        S: UnicodeString,
+        E: ToString,
     {
-        fn roundtrips(&self) {
-            let subject = self.subject.as_ref();
-            let dom = parse::<Utf16String>(subject).unwrap();
-            let output = restore_whitespace(&dom.to_html().to_string());
-            if output != subject {
-                AssertionFailure::from_spec(self)
-                    .with_expected(String::from(subject))
-                    .with_actual(output)
-                    .fail();
-            }
+        DomCreationError {
+            dom: Dom::new(vec![]),
+            parse_errors: vec![error.to_string()],
         }
     }
 
-    #[test]
-    fn parse_plain_text() {
-        assert_that!("some text").roundtrips();
+    enum Error {
+        NoBody,
+        UnknownNode(String),
     }
 
-    #[test]
-    fn parse_simple_tag() {
-        assert_that!("<strong>sdfds</strong>").roundtrips();
-    }
+    impl fmt::Display for Error {
+        fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+            match self {
+                Self::NoBody => {
+                    write!(
+                        formatter,
+                        "The `Document` does not have a `<body>` element"
+                    )
+                }
 
-    #[test]
-    fn parse_tag_with_surrounding_text() {
-        assert_that!("before <strong> within </strong> after").roundtrips();
-        assert_that!("before<strong>within</strong>after").roundtrips();
+                Self::UnknownNode(node_name) => {
+                    write!(formatter, "Node `{node_name}` is not supported")
+                }
+            }
+        }
     }
 
-    #[test]
-    fn parse_nested_tags() {
-        assert_that!("<b><em>ZZ</em></b>").roundtrips();
-        assert_that!("X<b>Y<em>ZZ</em>0</b>1").roundtrips();
-        assert_that!(" X <b> Y <em> ZZ </em> 0 </b> 1 ").roundtrips();
-    }
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use crate::{
+            tests::testutils_composer_model::restore_whitespace, ToHtml,
+        };
+        use wasm_bindgen_test::*;
+        use widestring::Utf16String;
 
-    #[test]
-    fn parse_tags_with_attributes() {
-        assert_that!(r#"<b><a href="http://example.com">ZZ</a></b>"#)
-            .roundtrips();
-    }
+        wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+
+        fn roundtrip(html: &str) {
+            let parse = parse::<Utf16String>(html);
+
+            assert!(
+                parse.is_ok(),
+                "Failed to parse the following HTML fragment: `{html}`"
+            );
+
+            let dom = parse.unwrap();
+            let html_again = restore_whitespace(&dom.to_html().to_string());
+
+            assert_eq!(html, html_again);
+        }
+
+        #[wasm_bindgen_test]
+        fn formatting() {
```

Comment:
I'd be really keen to share identical tests between the two features, if we can find a way.

Reply:
Yes, I would like to even have the same tests shared between both!